# Analysis of Listed Used Cars on Ebay
A dirtied scraped dataset of cars listings on ebay from Kaggle will be analysed and the functionality of pandas on Jupyter notebook will also be explored.
In the cell below, we will be importing the libraries needed for the analysis

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='cp437')

In [2]:
print(autos.info(), '\n')
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T▄V_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Exploring the dataset...

Exploring the file above, it is evident that some columns such as vehicleType, gearbox, model etc have null values and needs cleaning before proper analysis can be carried out. It was also observed that the columns have int64 or object types of element.

The column names format isn't also in the preffered python snakecase form. I will be converting it in the next cell.

In [3]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T▄V_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fueltype  \
0    manuell       158  andere  150,000km                   3      lpg   
1  automatik       286     7er  150,000km                   6   benzin   
2    manuell       102    golf   70,000km                   7   benzin   
3  automatik        71  fortwo   70,000km                   6   benzin   
4    manuell         0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## ...Changing column name

As evident in the cell above, the columns are now in the proper snakecase form.
In the next analysis, I will be exploring the dataset using the describe and value_counts methods.

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

## Dropping columns and changing data types 

From the description above, there are some columns like seller, offer_type, gear_box and no_of_pictures that does not provide variety of information and can be dropped.

The date_crawled also need to be converted to a datetime format. price and odometer are stored as strings instead of numeric forms.

In [5]:
autos.drop(columns=['offer_type', 'no_of_pictures'], inplace=True)

autos.rename({'odometer': 'odometer_km', 'price':'price_$'}, axis=1, inplace=True)

autos['odometer_km'] = autos['odometer_km'].str.replace(',' , '').str.replace('km' , '')
autos['price_$'] = autos['price_$'].str.replace(',', '').str.replace('$','')

autos['odometer_km'] = autos['odometer_km'].astype(int)
autos['price_$'] = autos['price_$'].astype(int)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T▄V_neu_ist_sehr_gepfleg...   

   seller  price_$   abtest vehicle_type  registration_year   gear_box  \
0  privat     5000  control          bus               2004    manuell   
1  privat     8500  control    limousine               1997  automatik   
2  privat     8990     test    limousine               2009    manuell   
3  privat     4350  control   kleinwagen               2007  automatik   
4  privat     1350     test        kombi               2003    manuell   

   power_ps   model  odometer_km  registration_month fueltype       brand  \
0       158  andere       150000                   3      lpg     peugeot   
1       286     7er       150000                   6   benzin         bmw   
2       102    golf        70000                   7   benzin  volkswagen   
3        71  fortwo        70000                   6   benzin       smart   
4         0   focus       150000                   7   benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Importing the datetime module 

The date_crawled, ad_created & last_seen are in string format and I will convert them to datetime object form in the cell below so that it will be more accessible.

In [6]:
from datetime import datetime


#converting date_crawled column
new_date = []
for obj in autos['date_crawled']:
    dt_obj = datetime.strptime(obj,'%Y-%m-%d %H:%M:%S')
    new_date.append(dt_obj)
autos['date_crawled'] = new_date

#converting ad_created column
new_date = []
for obj in autos['ad_created']:
    dt_obj = datetime.strptime(obj,'%Y-%m-%d %H:%M:%S')
    new_date.append(dt_obj)
autos['ad_created'] = new_date

#converting last_seen column
new_date = []
for obj in autos['last_seen']:
    dt_obj = datetime.strptime(obj,'%Y-%m-%d %H:%M:%S')
    new_date.append(dt_obj)
autos['last_seen'] = new_date

autos.head(7)

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T▄V_neu_ist_sehr_gepfleg...   
5 2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6 2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   

   seller  price_$   abtest vehicle_type  registration_year   gear_box  \
0  privat     5000  control          bus               2004    manuell   
1  privat     8500  control    limousine               1997  automatik   
2  privat     8990     test    limousine               2009    manuell   
3  privat     4350  control   kleinwagen               2007  automatik   
4  privat     1350     test        kombi               2003    manuell   
5  privat     7900     test          bus               2006  automatik   
6  privat      300     test    limousine               1995    manuell   

   power_ps    model  odometer_km  registration_month fueltype       brand  \
0       158   andere       150000                   3      lpg     peugeot   
1       286      7er       150000                   6   benzin         bmw   
2       102     golf        70000                   7   benzin  volkswagen   
3        71   fortwo        70000                   6   benzin       smart   
4         0    focus       150000                   7   benzin        ford   
5       150  voyager       150000                   4   diesel    chrysler   
6        90     golf       150000                   8   benzin  volkswagen   

  unrepaired_damage ad_created  postal_code           last_seen  
0              nein 2016-03-26        79588 2016-04-06 06:45:54  
1              nein 2016-04-04        71034 2016-04-06 14:45:08  
2              nein 2016-03-26        35394 2016-04-06 20:15:37  
3              nein 2016-03-12        33729 2016-03-15 03:16:28  
4              nein 2016-04-01        39218 2016-04-01 14:38:50  
5               NaN 2016-03-21        22962 2016-04-06 09:45:21  
6               NaN 2016-03-20        31535 2016-03-23 02:48:59

### Exploring odometer_km & price_$ columns

In the next cell, I will look for outliers in the two cells are remove them using boolean indexing method.

```
Based on the result of the fourth code cell, there are 2357 unique values in the price_$ column while the odometer has 17.
To get other descriptive analysis, I will run a describe method below
```

In [7]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [8]:
autos['price_$'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

In [9]:
autos['odometer_km'].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [10]:
autos['price_$'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_$, dtype: int64

In [11]:
autos['price_$'].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price_$, dtype: int64

### Removing rows with zero selling price

The last two preceeding showed that some listings are 0.0 which indicates that the listing is abnormal. Also, considering the mean price that is $9800 and the 25% and 75% from the descriptive analysis, I surmised that all the listing with 0.0 price should be removed.

In [12]:
autos = autos[autos['price_$'] > 0]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48579 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          48579 non-null datetime64[ns]
name                  48579 non-null object
seller                48579 non-null object
price_$               48579 non-null int32
abtest                48579 non-null object
vehicle_type          43990 non-null object
registration_year     48579 non-null int64
gear_box              46232 non-null object
power_ps              48579 non-null int64
model                 46116 non-null object
odometer_km           48579 non-null int32
registration_month    48579 non-null int64
fueltype              44544 non-null object
brand                 48579 non-null object
unrepaired_damage     39472 non-null object
ad_created            48579 non-null datetime64[ns]
postal_code           48579 non-null int64
last_seen             48579 non-null datetime64[ns]
dtypes: datetime64[ns](3), int32(2), int64(4), object(9)
memory usage: 6.7+ MB


## Manipulating dates

I have done the conversion of date_crawled, ad_created & last_seen to datetime object in preceding cells. Below, I will be exploring the registration year column to test for outliers

In [13]:
autos['registration_year'].describe()

count    48579.000000
mean      2004.753000
std         88.631663
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Both the minimum value of 1000 and the maximum value of 9999 are unrealistic as a car cannont be registered since that far from the present. 
Therefore, any car registered in dates above 2016 or below 1950 should be eliminated (after ensuring that the removed part does not make up for a significant part of the dataset). 

In [14]:
year_bool = (autos['registration_year'] >= 1950) & (autos['registration_year'] <= 2016)

autos = autos[year_bool]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46673 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          46673 non-null datetime64[ns]
name                  46673 non-null object
seller                46673 non-null object
price_$               46673 non-null int32
abtest                46673 non-null object
vehicle_type          43972 non-null object
registration_year     46673 non-null int64
gear_box              44568 non-null object
power_ps              46673 non-null int64
model                 44485 non-null object
odometer_km           46673 non-null int32
registration_month    46673 non-null int64
fueltype              43359 non-null object
brand                 46673 non-null object
unrepaired_damage     38369 non-null object
ad_created            46673 non-null datetime64[ns]
postal_code           46673 non-null int64
last_seen             46673 non-null datetime64[ns]
dtypes: datetime64[ns](3), int32(2), int64(4), object(9)
memory usage: 6.4+ MB


The number of dismissed columns when designating 1950 and 2016 as the lower and upper limit is a tiny percent of the whole dataset (3%). 

In [15]:
autos['registration_year'].value_counts(normalize=True).head()

2000    0.067619
2005    0.062906
1999    0.062113
2004    0.057914
2003    0.057828
Name: registration_year, dtype: float64

## Aggregating the brand column

I will use the aggregating method to create a dictionary containing brands with their mean price.

In [27]:
autos['brand'].value_counts()

volkswagen        9864
bmw               5137
opel              5020
mercedes_benz     4502
audi              4041
ford              3262
renault           2200
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            655
toyota             593
hyundai            468
sonstige_autos     453
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             64
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

I will be aggregating with the top 20 brands and determining their average price since they are the popular ones. 

In [36]:
brand_dict = {}

for brand in autos['brand'].value_counts().head(20).index:
    rows_brand = autos[autos['brand'] == brand]
    mean_price = rows_brand['price_$'].sum() / rows_brand.shape[0]
    brand_dict[brand] = mean_price
    #price = 0
    #count = 0
    #for row in autos:
        #if row[5] == brand:
            #count += 1
            #price += auto['price_$']
    #mean_price = price / count
    #brand_dict[brand] = mean_price
    
brand_dict

{'volkswagen': 6726.254055150041,
 'bmw': 8570.910064239828,
 'opel': 5434.952788844622,
 'mercedes_benz': 30835.675477565528,
 'audi': 9336.687453600594,
 'ford': 7447.770999386879,
 'renault': 2475.7172727272728,
 'peugeot': 3094.0172290021537,
 'fiat': 2813.748538011696,
 'seat': 4397.230949589683,
 'skoda': 6368.0,
 'nissan': 4743.40252454418,
 'mazda': 4112.596614950635,
 'smart': 3580.2239031770046,
 'citroen': 45486.68549618321,
 'toyota': 5167.091062394604,
 'hyundai': 5365.254273504273,
 'sonstige_autos': 45786.89845474614,
 'volvo': 4946.501170960188,
 'mini': 10613.459657701711}

The top six brands varies in their average price as the dictionary above shows. The Mercedes and BMW are somewhat more expensive than the others.
To check whether cost is related to the brand mileage, I will be aggregating base on the odometer_km column as well and they compare them.

## Aggregating the odometer_km column

In the next few cells, I will aggregate the odometer_column to compare aggregated mean prices with the mileage. 

In [31]:
odometer_dict = {}

for brand in autos['brand'].value_counts().head(20).index:
    rows_brand = autos[autos['brand'] == brand]
    mean_price = rows_brand['odometer_km'].sum() / rows_brand.shape[0]
    odometer_dict[brand] = mean_price
    
odometer_dict

{'volkswagen': 128716.54501216544,
 'bmw': 132600.73973136072,
 'opel': 129340.63745019921,
 'mercedes_benz': 130836.29498000888,
 'audi': 129157.38678544914,
 'ford': 124351.62477007971,
 'renault': 128127.27272727272,
 'peugeot': 127153.62526920316,
 'fiat': 117121.9715956558,
 'seat': 121131.30128956624,
 'skoda': 110848.5639686684,
 'nissan': 118330.99579242637,
 'mazda': 124464.03385049365,
 'smart': 99326.77760968229,
 'citroen': 119572.51908396947,
 'toyota': 115944.35075885328,
 'hyundai': 106442.30769230769,
 'sonstige_autos': 90397.35099337748,
 'volvo': 138067.9156908665,
 'mini': 88105.13447432763}

```
Now, I will convert both dictionary to series object to later join them together to form a dataframe.
```

In [37]:
sr_price = pd.Series(brand_dict)
sr_odometer = pd.Series(odometer_dict)

#converting to Dataframe
df = pd.DataFrame(sr_price, columns=['mean_price'])
df['mileage'] = sr_odometer

df

mean_price        mileage
volkswagen       6726.254055  128716.545012
bmw              8570.910064  132600.739731
opel             5434.952789  129340.637450
mercedes_benz   30835.675478  130836.294980
audi             9336.687454  129157.386785
ford             7447.770999  124351.624770
renault          2475.717273  128127.272727
peugeot          3094.017229  127153.625269
fiat             2813.748538  117121.971596
seat             4397.230950  121131.301290
skoda            6368.000000  110848.563969
nissan           4743.402525  118330.995792
mazda            4112.596615  124464.033850
smart            3580.223903   99326.777610
citroen         45486.685496  119572.519084
toyota           5167.091062  115944.350759
hyundai          5365.254274  106442.307692
sonstige_autos  45786.898455   90397.350993
volvo            4946.501171  138067.915691
mini            10613.459658   88105.134474

# Conclusion

The dataframe above shows the brand mean price is not indicative of the average mileage travelled by the brands. Volvo with an average price of 4,946 has almost the same mileage value with Mercedez Benz with an average price of 30,835. 


Thank you!